In [2]:
# IMPORT
import numpy as np
import pandas as pd
from pandas import datetime

import matplotlib.pyplot as plt
import seaborn as sns # advanced vizs
%matplotlib inline



In [3]:
import statistics
import math

In [4]:
from bokeh.plotting import figure, output_file, show
from bokeh.layouts import gridplot

In [5]:
store_typeA = pd.read_csv('./ModelA/store_typeA.csv',
                    parse_dates = True,
                    low_memory = False,
                    index_col = 'Date')

store_typeA.head()

Store  DayOfWeek  Sales  Customers  Open  Promo StateHoliday  \
Date                                                                       
2013-01-01    530          2   2907        532     1      0            a   
2013-01-02    773          3   5955        864     1      0            0   
2013-01-02    712          3   6377        615     1      0            0   
2013-01-02    929          3   5011        663     1      0            0   
2013-01-02    276          3   3406        538     1      0            0   

            SchoolHoliday StoreType Assortment  CompetitionDistance  \
Date                                                                  
2013-01-01              1         a          c              18160.0   
2013-01-02              1         a          a                200.0   
2013-01-02              1         a          a               4870.0   
2013-01-02              1         a          c               4820.0   
2013-01-02              1         a          a               2960.0   

            CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
Date                                                                      
2013-01-01                        0.0                       0.0       0   
2013-01-02                        0.0                       0.0       0   
2013-01-02                        9.0                    2007.0       1   
2013-01-02                        9.0                    2013.0       0   
2013-01-02                       10.0                    2014.0       1   

            Promo2SinceWeek  Promo2SinceYear     PromoInterval  
Date                                                            
2013-01-01              0.0              0.0                 0  
2013-01-02              0.0              0.0                 0  
2013-01-02             45.0           2009.0   Jan,Apr,Jul,Oct  
2013-01-02              0.0              0.0                 0  
2013-01-02             36.0           2013.0  Mar,Jun,Sept,Dec

In [22]:
store_typeA.shape

# store_typeA_avgSales = store_typeA.groupby('Store').mean()
# store_typeA_avgSales = store_typeA_avgSales.drop(['DayOfWeek','Customers','Open','Promo','SchoolHoliday','CompetitionDistance','CompetitionOpenSinceMonth','CompetitionOpenSinceYear','Promo2','Promo2SinceWeek','Promo2SinceYear'],axis=1)
# store_typeA_avgSales.head()

(457042, 17)

In [30]:
len(set(store_typeA.Store))

602

In [6]:
train_A = pd.read_csv('./ModelA/train_A.csv',
                     parse_dates = True,
                    low_memory = False,
                    index_col = 'Date')
train_A.head()

Store  DayOfWeek  Sales  Customers  Open  Promo StateHoliday  \
Date                                                                       
2013-01-02    480          3   4827        660     1      0            0   
2013-01-03    480          4   4354        614     1      0            0   
2013-01-04    480          5   4027        560     1      0            0   
2013-01-05    480          6   2662        367     1      0            0   
2013-01-07    480          1   8135        936     1      1            0   

            SchoolHoliday StoreType Assortment  CompetitionDistance  \
Date                                                                  
2013-01-02              1         a          a                300.0   
2013-01-03              1         a          a                300.0   
2013-01-04              1         a          a                300.0   
2013-01-05              0         a          a                300.0   
2013-01-07              0         a          a                300.0   

            CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
Date                                                                      
2013-01-02                       11.0                    2012.0       0   
2013-01-03                       11.0                    2012.0       0   
2013-01-04                       11.0                    2012.0       0   
2013-01-05                       11.0                    2012.0       0   
2013-01-07                       11.0                    2012.0       0   

            Promo2SinceWeek  Promo2SinceYear PromoInterval  
Date                                                        
2013-01-02              0.0              0.0             0  
2013-01-03              0.0              0.0             0  
2013-01-04              0.0              0.0             0  
2013-01-05              0.0              0.0             0  
2013-01-07              0.0              0.0             0

In [21]:
train_A.shape

(45159, 17)

In [29]:
len(set(train_A.Store))

60

# Sample as a representation of population
10% of stores in store type A selected in sample(60 out of 602 stores)

In [7]:
final_A_10pct_sales = train_A.groupby(train_A.index).Sales.mean().reset_index()
final_storeA_sales = store_typeA.groupby(store_typeA.index).Sales.mean().reset_index()

In [8]:
# Daily sales bad graph

output_file("./ModelA/sales_daily.html")

p = figure(plot_width=2000, plot_height=500, x_axis_type="datetime")
p.legend.location = "top_left"
p.title.text = "Daily average Sales 2013-2015"
p.title.align = "left"
p.title.text_font_size = "25px"


# add a line renderer
p.line(final_storeA_sales['Date'], final_storeA_sales['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
p.line(final_A_10pct_sales['Date'],final_A_10pct_sales['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')
show(p)

/anaconda3/lib/python3.6/site-packages/bokeh/models/plots.py:726: UserWarning: 
You are attemptings to set `plot.legend.location` on a plot that has zero legends added, this will have no effect.

Before legend properties can be set, you must add a Legend explicitly, or call a glyph method with the 'legend' parameter set.

  warnings.warn(_LEGEND_EMPTY_WARNING % attr)


In [9]:
final_A_10pct_sales['Year'] = [y.year for y in final_A_10pct_sales['Date']] 
final_A_10pct_sales['WeekofYear'] = [y.isocalendar()[1] for y in final_A_10pct_sales['Date']]

final_storeA_sales['Year'] = [y.year for y in final_storeA_sales['Date']] 
final_storeA_sales['WeekofYear'] = [y.isocalendar()[1] for y in final_storeA_sales['Date']]

# to plot yearly average sales(daily) to see underlying trend
final_A_10pct_sales_2013 = final_A_10pct_sales[final_A_10pct_sales['Year']==2013]
final_storeA_sales_2013= final_storeA_sales[final_storeA_sales['Year']==2013]

final_A_10pct_sales_2014 = final_A_10pct_sales[final_A_10pct_sales['Year']==2014]
final_storeA_sales_2014 = final_storeA_sales[final_storeA_sales['Year']==2014]

final_A_10pct_sales_2015 = final_A_10pct_sales[final_A_10pct_sales['Year']==2015]
final_storeA_sales_2015 = final_storeA_sales[final_storeA_sales['Year']==2015]

In [10]:
output_file("./ModelA/sales_daily_yearwise.html")

s1 = figure(plot_width=600, plot_height=400,title='2013 daily average sales',x_axis_type="datetime")

# add a line renderer
s1.line(final_A_10pct_sales_2013['Date'], final_A_10pct_sales_2013['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
s1.line(final_storeA_sales_2013['Date'],final_storeA_sales_2013['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')

s1.legend.location = "top_left"

s2 = figure(plot_width=600, plot_height=400,title='2014 daily average sales',x_axis_type="datetime")

s2.line(final_A_10pct_sales_2014['Date'], final_A_10pct_sales_2014['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
s2.line(final_storeA_sales_2014['Date'],final_storeA_sales_2014['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')

s2.legend.location = "top_left"

s3 = figure(plot_width=600, plot_height=400,title='2015 daily average sales',x_axis_type="datetime")
s3.line(final_A_10pct_sales_2015['Date'], final_A_10pct_sales_2015['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
s3.line(final_storeA_sales_2015['Date'],final_storeA_sales_2015['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')


s3.legend.location = "top_left"

p2 = gridplot([[s1, s2, s3]])


# s1.title.text = "Daily average Sales yearwise"
# s1.title.align = "left"
# s1.title.text_font_size = "25px"

show(p2)

In [11]:
final_A_10pct_sales_2013_weekly = final_A_10pct_sales_2013.groupby(['WeekofYear']).mean()
final_storeA_sales_2013_weekly = final_storeA_sales_2013.groupby(['WeekofYear']).mean()

final_A_10pct_sales_2014_weekly = final_A_10pct_sales_2014.groupby(['WeekofYear']).mean()
final_storeA_sales_2014_weekly = final_storeA_sales_2014.groupby(['WeekofYear']).mean()

final_A_10pct_sales_2015_weekly = final_A_10pct_sales_2015.groupby(['WeekofYear']).mean()
final_storeA_sales_2015_weekly = final_storeA_sales_2015.groupby(['WeekofYear']).mean()

In [12]:

# average weekly sales

output_file("./ModelA/sales_weekly_yearwise.html")

r1 = figure(plot_width=400, plot_height=400,title='2013 weekly average sales',x_axis_type="datetime")

# add a line renderer
r1.line(final_A_10pct_sales_2013_weekly.index, final_A_10pct_sales_2013_weekly['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
r1.line(final_storeA_sales_2013_weekly.index,final_storeA_sales_2013_weekly['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')

r1.legend.location = "top_left"


r2 = figure(plot_width=400, plot_height=400,title='2014 weekly average sales',x_axis_type="datetime",)

r2.line(final_A_10pct_sales_2014_weekly.index, final_A_10pct_sales_2014_weekly['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
r2.line(final_storeA_sales_2014_weekly.index,final_storeA_sales_2014_weekly['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')

r2.legend.location = "top_left"

r3 = figure(plot_width=400, plot_height=400,title='2015 weekly average sales',x_axis_type="datetime")

r3.line(final_A_10pct_sales_2015_weekly.index, final_A_10pct_sales_2015_weekly['Sales'], line_width=2,line_color = 'red',legend='Sampled stores')
r3.line(final_storeA_sales_2015_weekly.index,final_storeA_sales_2015_weekly['Sales'], line_width=2,line_color = 'blue',legend='All stores type A')
r3.legend.location = "top_left"



p3 = gridplot([[r1, r2, r3]])



show(p3)

# Modelling

In [13]:
X = train_A.copy()


In [14]:
X = X.drop(['StoreType','Promo2','Promo2SinceWeek','Promo2SinceYear','PromoInterval','Sales','CompetitionOpenSinceMonth','CompetitionOpenSinceYear'], axis =1)

X.columns

Index(['Store', 'DayOfWeek', 'Customers', 'Open', 'Promo', 'StateHoliday',
       'SchoolHoliday', 'Assortment', 'CompetitionDistance'],
      dtype='object')

In [80]:
X.columns

Index(['Store', 'DayOfWeek', 'Customers', 'Open', 'Promo', 'StateHoliday',
       'SchoolHoliday', 'Assortment', 'CompetitionDistance'],
      dtype='object')

In [ ]:
# from sklearn.preprocessing import LabelEncoder, OneHotEncoder
# le_state_holiday = LabelEncoder()
# X['StateHoliday'] = le_state_holiday.fit_transform(X['StateHoliday'])
# # onc_state_holiday = OneHotEncoder()


# list(le_state_holiday.classes_)
#['0', 'a']

NOT VALID NOW as 4 classes are available
Since there is only one type of holiday, it does not make sense to keep using label encoder, we can simply convert all state holiday to 1 (as using label encoder converts all a to 0, but 0 indicates not holiday)

In [15]:
map_holiday = {'a':1,'b':2, 'c':3,'d':4, '0':0}

X['StateHoliday'] = X['StateHoliday'].map(map_holiday)

set(X['StateHoliday'].values)

{0, 1, 2}

In [82]:
X.groupby(['StateHoliday']).count()

Store  DayOfWeek  Customers   Open  Promo  SchoolHoliday  \
StateHoliday                                                             
0             45124      45124      45124  45124  45124          45124   
1                29         29         29     29     29             29   
2                 6          6          6      6      6              6   

              Assortment  CompetitionDistance  
StateHoliday                                   
0                  45124                45124  
1                     29                   29  
2                      6                    6

In [16]:
X = pd.concat([X,pd.get_dummies(X['StateHoliday'],prefix=['StateHoliday_type'],drop_first=True)],axis=1)

X = X.drop(['StateHoliday'],axis =1)

In [85]:
X.head()

Store  DayOfWeek  Customers  Open  Promo  SchoolHoliday  \
Date                                                                  
2013-01-02    480          3        660     1      0              1   
2013-01-03    480          4        614     1      0              1   
2013-01-04    480          5        560     1      0              1   
2013-01-05    480          6        367     1      0              0   
2013-01-07    480          1        936     1      1              0   

           Assortment  CompetitionDistance  ['StateHoliday_type']_1  \
Date                                                                  
2013-01-02          a                300.0                        0   
2013-01-03          a                300.0                        0   
2013-01-04          a                300.0                        0   
2013-01-05          a                300.0                        0   
2013-01-07          a                300.0                        0   

            ['StateHoliday_type']_2  
Date                                 
2013-01-02                        0  
2013-01-03                        0  
2013-01-04                        0  
2013-01-05                        0  
2013-01-07                        0

In [ ]:
# from sklearn.preprocessing import LabelEncoder, OneHotEncoder
# onc_state_holiday = OneHotEncoder()
# X['StateHoliday'] = onc_state_holiday.fit_transform(X['StateHoliday'].values.reshape(-1,1))
# # encoded_stateholiday = [i[0] for i in encoded_stateholiday]
# encoded_stateholiday



Before transforming assortment, we check which types are available

In [17]:
X.groupby(['Assortment']).count()

Store  DayOfWeek  Customers   Open  Promo  SchoolHoliday  \
Assortment                                                             
a           25712      25712      25712  25712  25712          25712   
c           19447      19447      19447  19447  19447          19447   

            CompetitionDistance  ['StateHoliday_type']_1  \
Assortment                                                 
a                         25712                    25712   
c                         19447                    19447   

            ['StateHoliday_type']_2  
Assortment                           
a                             25712  
c                             19447

since only 2 labels are present, we can use maps again

In [18]:
map_assortment = {'a':0,'c':1}

X['Assortment'] = X['Assortment'].map(map_assortment) 

set(X['Assortment'].values)

y = train_A['Sales'].values.reshape(-1,1)

# store date index before scaling
train_size = round(0.80*len(X))
X_train_dates = X.index[:train_size]
X_test_dates = X.index[train_size:]
y_train_dates = X.index[:train_size]
y_test_dates = X.index[train_size:]

from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X=sc.fit_transform(X)
sc_y = StandardScaler()
y = sc_y.fit_transform(y)

/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)
/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [22]:
len(X)==len(y)

True

In [31]:
# train_size = round(0.80*len(X))

In [32]:
# train_size

36542

In [19]:
X_train = X[:train_size,:].copy()
X_test = X[train_size:,:].copy()
y_train = y[:train_size].copy()
y_test = y[train_size:].copy()

In [34]:
X_train

array([[-1.15085681,  0.85999576, -0.07648864, ...,  0.68599602,
        -0.02144648, -0.01146174],
       [-1.15085681,  0.27950625, -0.15893284, ...,  0.68599602,
        -0.02144648, -0.01146174],
       [-1.15085681, -0.30098327, -0.14019552, ...,  0.68599602,
        -0.02144648, -0.01146174],
       ...,
       [ 0.88096629, -1.46196229, -0.64797681, ...,  0.67958819,
        -0.02144648, -0.01146174],
       [ 0.88096629,  1.44048527, -1.06394525, ...,  0.67958819,
        -0.02144648, -0.01146174],
       [ 0.88096629,  0.85999576, -0.64048189, ...,  0.67958819,
        -0.02144648, -0.01146174]])

In [20]:
# from sklearn.model_selection import train_test_split
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)



from sklearn.svm import SVR

In [21]:
svm_reg = SVR()
svm_reg.fit(X_train, y_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/anaconda3/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1,
  gamma='auto_deprecated', kernel='rbf', max_iter=-1, shrinking=True,
  tol=0.001, verbose=False)

In [22]:
y_pred_svm = svm_reg.predict(X_test)

In [23]:
from sklearn.metrics import mean_squared_error
np.sqrt(mean_squared_error(y_test, y_pred_svm))

0.39539908877910585

In [24]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred_svm)

0.9135729321521766

In [28]:
y_pred_svm.shape

(9053,)

In [29]:
# y_test1 = [y1[0] for y1 in y_test] 

In [ ]:
# y_pred_svm=y_pred_svm.reshape(-1,1)

# Visualisation
---
## Predictions

In [25]:
y_test_inverse_list = [a[0] for a in sc_y.inverse_transform(y_test)]
y_pred_inverse_list = [b for b in sc_y.inverse_transform(y_pred_svm)]

In [26]:
results = pd.DataFrame(list(zip(y_test_inverse_list, y_pred_inverse_list)),columns=['Y_test','Y_Predicted'],index=y_test_dates)

In [27]:
results=results.sort_index()

In [ ]:
results

In [28]:
# Daily data
output_file("./ModelA/predictions.html")

p4 = figure(plot_width=1000, plot_height=700, x_axis_type="datetime")

p4 = figure(plot_width=1000, plot_height=700, x_axis_type="datetime")
p4.title.text = "Daily Sales Predictions forecasting"
p4.title.align = "left"
p4.title.text_font_size = "25px"
# add a line renderer
p4.line(results.index, results['Y_test'], line_width=2,line_color = 'red', legend = 'Test set')
p4.line(results.index,results['Y_Predicted'], line_width=2,line_color = 'blue', legend = 'Prediction set')
show(p4)

In [30]:
# results_weekly = results.resample('W').agg({'Y_test':'mean','Y_Predicted':'mean'})
results['RM_ytest']=results['Y_test'].rolling(7).mean()
results['RM_ypred']=results['Y_Predicted'].rolling(7).mean()

In [31]:
# Daily rolling(7) data
output_file("./ModelA/rolling_predictions_daily.html")

p5 = figure(plot_width=3000, plot_height=700, x_axis_type="datetime")
p5.title.text = "Daily rolling Sales forecasting"
p5.title.align = "left"
p5.title.text_font_size = "25px"
# add a line renderer
p5.line(results.index, results['RM_ytest'], line_width=2,line_color = 'red',legend='Test set')
p5.line(results.index,results['RM_ypred'], line_width=2,line_color = 'blue',legend='Prediction set')
show(p5)


In [122]:
results_weekly = results.resample('W').agg({'Y_test':'mean','Y_Predicted':'mean'})
results_weekly['RM_ytest']=results_weekly['Y_test'].rolling(5).mean()
results_weekly['RM_ypred']=results_weekly['Y_Predicted'].rolling(5).mean()

In [120]:
results.shape

(9032, 4)

In [121]:
results_weekly.shape

(135, 4)

In [126]:
# weekly data rolling(5) --> next week's prediction depends on last month's weekly predictions
output_file("./ModelA/rolling_predictions_weekly.html")

p6 = figure(plot_width=1500, plot_height=700, x_axis_type="datetime")


p6.title.text = "Downsampled weekly rolled(5) average Sales 2013-2015"
p6.title.align = "left"
p6.title.text_font_size = "25px"

# add a line renderer
p6.line(results_weekly.index, results_weekly['RM_ytest'], line_width=2,line_color = 'red',legend='Sampled stores')
p6.line(results_weekly.index,results_weekly['RM_ypred'], line_width=2,line_color = 'blue',legend='All stores type A')

p6.legend.location = "top_left"
show(p6)

In [54]:
final_A_10pct_sales = train_A.groupby(train_A.index).Sales.mean().reset_index()
final_storeA_sales = store_typeA.groupby(store_typeA.index).Sales.mean().reset_index()

In [55]:
final_A_10pct_sales

Date         Sales
0   2013-01-02   6528.333333
1   2013-01-03   5965.150000
2   2013-01-04   6010.016667
3   2013-01-05   5336.266667
4   2013-01-07  10551.633333
5   2013-01-08   8589.466667
6   2013-01-09   7380.833333
7   2013-01-10   7515.283333
8   2013-01-11   7494.783333
9   2013-01-12   5346.083333
10  2013-01-14   5783.633333
11  2013-01-15   5395.800000
12  2013-01-16   5178.033333
13  2013-01-17   5260.333333
14  2013-01-18   5593.300000
15  2013-01-19   5218.550000
16  2013-01-21   8516.633333
17  2013-01-22   8118.716667
18  2013-01-23   7392.850000
19  2013-01-24   7117.133333
20  2013-01-25   7200.650000
21  2013-01-26   5087.400000
22  2013-01-28   5694.166667
23  2013-01-29   5276.133333
24  2013-01-30   5642.850000
25  2013-01-31   6926.416667
26  2013-02-01   6564.383333
27  2013-02-02   6065.233333
28  2013-02-04  10044.716667
29  2013-02-05   8560.316667
..         ...           ...
754 2015-06-27   5525.716667
755 2015-06-29  11007.200000
756 2015-06-30  10890.466667
757 2015-07-01   8407.216667
758 2015-07-02   7762.593220
759 2015-07-03   7484.101695
760 2015-07-04   5152.372881
761 2015-07-06   7303.152542
762 2015-07-07   6175.711864
763 2015-07-08   6348.525424
764 2015-07-09   6420.576271
765 2015-07-10   6279.169492
766 2015-07-11   5315.338983
767 2015-07-13   9735.406780
768 2015-07-14   8590.949153
769 2015-07-15   8083.796610
770 2015-07-16   7868.965517
771 2015-07-17   7004.862069
772 2015-07-18   5627.101695
773 2015-07-20   6949.847458
774 2015-07-21   6040.677966
775 2015-07-22   5730.169492
776 2015-07-23   6206.593220
777 2015-07-24   5956.440678
778 2015-07-25   4987.932203
779 2015-07-27   9547.423729
780 2015-07-28   8340.677966
781 2015-07-29   7798.661017
782 2015-07-30   8119.406780
783 2015-07-31   9317.779661

[784 rows x 2 columns]

## Sampling graphs
daily sales 2014 -->sample(blue) vs population(red)

In [59]:


output_file("sales_daily.html")

p = figure(plot_width=1000, plot_height=1000, x_axis_type="datetime")

# add a line renderer
p.line(final_storeA_sales['Date'], final_storeA_sales['Sales'], line_width=2,line_color = 'red')
p.line(final_A_10pct_sales['Date'],final_A_10pct_sales['Sales'], line_width=2,line_color = 'blue')
show(p)

In [63]:
final_A_10pct_sales['Year'] = [y.year for y in final_A_10pct_sales['Date']] 
final_A_10pct_sales['WeekofYear'] = [y.isocalendar()[1] for y in final_A_10pct_sales['Date']]

final_storeA_sales['Year'] = [y.year for y in final_storeA_sales['Date']] 
final_storeA_sales['WeekofYear'] = [y.isocalendar()[1] for y in final_storeA_sales['Date']]

final_A_10pct_sales_2014 = final_A_10pct_sales[final_A_10pct_sales['Year']==2014]
final_storeA_sales_2014 = final_storeA_sales[final_storeA_sales['Year']==2014]

In [64]:
output_file("sales_daily2014.html")

p = figure(plot_width=1000, plot_height=1000)

# add a line renderer
p.line(final_A_10pct_sales_2014['Date'], final_A_10pct_sales_2014['Sales'], line_width=2,line_color = 'red')
p.line(final_storeA_sales_2014['Date'],final_storeA_sales_2014['Sales'], line_width=2,line_color = 'blue')
show(p)

In [65]:
final_A_10pct_sales_2014_weekly = final_A_10pct_sales_2014.groupby(['WeekofYear']).mean()
final_storeA_sales_2014_weekly = final_storeA_sales_2014.groupby(['WeekofYear']).mean()

In [67]:
final_A_10pct_sales_2014_weekly

Sales  Year
WeekofYear                    
1            6350.378431  2014
2            8199.278986  2014
3            5585.892750  2014
4            7575.641243  2014
5            5980.827684  2014
6            8242.285311  2014
7            5875.098870  2014
8            7638.369397  2014
9            5966.911111  2014
10           8137.299405  2014
11           5654.516667  2014
12           7350.319444  2014
13           5740.683333  2014
14           8200.191667  2014
15           5977.266667  2014
16           9283.630000  2014
17           5676.970000  2014
18           9040.856667  2014
19           8041.505556  2014
20           5720.527778  2014
21           7524.330556  2014
22           6696.664520  2014
23           8413.064972  2014
24           6252.827119  2014
25           8104.568785  2014
26           5826.121092  2014
27           8523.289542  2014
28           5921.316993  2014
29           7450.903595  2014
30           5561.793333  2014
31           7973.890719  2014
32           7452.470588  2014
33           5667.901961  2014
34           7462.343137  2014
35           6332.127451  2014
36           7988.956013  2014
37           5684.973469  2014
38           7188.926871  2014
39           5590.537124  2014
40           8506.682353  2014
41           7562.986928  2014
42           5549.307190  2014
43           6840.160131  2014
44           6324.039476  2014
45           8235.107843  2014
46           7396.588235  2014
47           6061.796364  2014
48           8803.055556  2014
49           9333.398693  2014
50           6850.421569  2014
51          10393.826797  2014
52           8179.426471  2014

In [68]:
output_file("sales_weekly2014.html")

p = figure(plot_width=1000, plot_height=1000)

# add a line renderer
p.line(final_A_10pct_sales_2014_weekly.index, final_A_10pct_sales_2014_weekly['Sales'], line_width=2,line_color = 'red')
p.line(final_storeA_sales_2014_weekly.index,final_storeA_sales_2014_weekly['Sales'], line_width=2,line_color = 'blue')
show(p)

In [ ]:
---

In [ ]:
# from bokeh.plotting import figure, output_file, show

# output_file("sales.html")

# p = figure(plot_width=1000, plot_height=1000)

# # add a line renderer
# p.line(final_storeA_sales['Date'], final_storeA_sales['Sales'], line_width=2,line_color = 'red')
# p.line(final_A_10pct_sales['Date'],final_A_10pct_sales['Sales'], line_width=2,line_color = 'blue')
# show(p)

train_A is top 10% representative sample

In [283]:
final_A_10pct_sales.shape

(784, 2)

In [284]:
final_A_10pct_sales

Date   Sales
0   2013-01-02  348256
1   2013-01-03  319927
2   2013-01-04  325878
3   2013-01-05  261786
4   2013-01-07  565617
5   2013-01-08  454779
6   2013-01-09  385230
7   2013-01-10  393174
8   2013-01-11  398019
9   2013-01-12  263100
10  2013-01-14  301849
11  2013-01-15  284524
12  2013-01-16  267723
13  2013-01-17  278913
14  2013-01-18  303421
15  2013-01-19  264462
16  2013-01-21  461199
17  2013-01-22  430151
18  2013-01-23  394720
19  2013-01-24  372113
20  2013-01-25  381217
21  2013-01-26  254118
22  2013-01-28  303622
23  2013-01-29  277737
24  2013-01-30  297128
25  2013-01-31  355710
26  2013-02-01  353467
27  2013-02-02  292361
28  2013-02-04  530477
29  2013-02-05  456549
..         ...     ...
754 2015-06-27  283942
755 2015-06-29  599316
756 2015-06-30  579963
757 2015-07-01  451934
758 2015-07-02  415096
759 2015-07-03  406832
760 2015-07-04  271142
761 2015-07-06  375976
762 2015-07-07  317955
763 2015-07-08  337809
764 2015-07-09  342422
765 2015-07-10  348722
766 2015-07-11  281724
767 2015-07-13  534345
768 2015-07-14  467260
769 2015-07-15  437324
770 2015-07-16  418844
771 2015-07-17  386057
772 2015-07-18  284547
773 2015-07-20  364457
774 2015-07-21  318528
775 2015-07-22  303055
776 2015-07-23  325797
777 2015-07-24  319536
778 2015-07-25  258161
779 2015-07-27  522316
780 2015-07-28  449444
781 2015-07-29  413437
782 2015-07-30  430056
783 2015-07-31  493077

[784 rows x 2 columns]

In [306]:
# final_A_10pct_sales.loc[0,'Date'].isocalendar()[1]

1

In [286]:
final_storeA_sales.shape

(916, 2)

In [287]:
final_storeA_sales

Date    Sales
0   2013-01-01     2907
1   2013-01-02  3920386
2   2013-01-03  3528027
3   2013-01-04  3610215
4   2013-01-05  2964587
5   2013-01-06     6254
6   2013-01-07  6093933
7   2013-01-08  4952478
8   2013-01-09  4297272
9   2013-01-10  4314074
10  2013-01-11  4389696
11  2013-01-12  3004666
12  2013-01-13     2790
13  2013-01-14  3346808
14  2013-01-15  3163977
15  2013-01-16  3012749
16  2013-01-17  3072432
17  2013-01-18  3345679
18  2013-01-19  2954399
19  2013-01-20     3229
20  2013-01-21  4940752
21  2013-01-22  4680093
22  2013-01-23  4293760
23  2013-01-24  4092087
24  2013-01-25  4192388
25  2013-01-26  2888014
26  2013-01-27     3205
27  2013-01-28  3374887
28  2013-01-29  3087306
29  2013-01-30  3309817
..         ...      ...
886 2015-07-02  4565168
887 2015-07-03  4491161
888 2015-07-04  2952626
889 2015-07-05    40384
890 2015-07-06  4137007
891 2015-07-07  3551296
892 2015-07-08  3730150
893 2015-07-09  3735672
894 2015-07-10  3775550
895 2015-07-11  3070374
896 2015-07-12    43187
897 2015-07-13  5753861
898 2015-07-14  5059171
899 2015-07-15  4779650
900 2015-07-16  4535275
901 2015-07-17  4164812
902 2015-07-18  3157657
903 2015-07-19    55404
904 2015-07-20  3996292
905 2015-07-21  3515258
906 2015-07-22  3337937
907 2015-07-23  3594960
908 2015-07-24  3499777
909 2015-07-25  2847063
910 2015-07-26    55506
911 2015-07-27  5603824
912 2015-07-28  4910264
913 2015-07-29  4524391
914 2015-07-30  4685380
915 2015-07-31  5380356

[916 rows x 2 columns]

In [279]:
len(final_A_10pct_sales)

784

In [292]:
final_storeA_sales.loc[1,'Date']==final_A_10pct_sales.loc[0,'Date']

True

In [291]:
final_A_10pct_sales.loc[0,'Date']

Timestamp('2013-01-02 00:00:00')

In [277]:
final_storeA_sales

Date    Sales
0   2013-01-01     2907
1   2013-01-02  3920386
2   2013-01-03  3528027
3   2013-01-04  3610215
4   2013-01-05  2964587
5   2013-01-06     6254
6   2013-01-07  6093933
7   2013-01-08  4952478
8   2013-01-09  4297272
9   2013-01-10  4314074
10  2013-01-11  4389696
11  2013-01-12  3004666
12  2013-01-13     2790
13  2013-01-14  3346808
14  2013-01-15  3163977
15  2013-01-16  3012749
16  2013-01-17  3072432
17  2013-01-18  3345679
18  2013-01-19  2954399
19  2013-01-20     3229
20  2013-01-21  4940752
21  2013-01-22  4680093
22  2013-01-23  4293760
23  2013-01-24  4092087
24  2013-01-25  4192388
25  2013-01-26  2888014
26  2013-01-27     3205
27  2013-01-28  3374887
28  2013-01-29  3087306
29  2013-01-30  3309817
..         ...      ...
886 2015-07-02  4565168
887 2015-07-03  4491161
888 2015-07-04  2952626
889 2015-07-05    40384
890 2015-07-06  4137007
891 2015-07-07  3551296
892 2015-07-08  3730150
893 2015-07-09  3735672
894 2015-07-10  3775550
895 2015-07-11  3070374
896 2015-07-12    43187
897 2015-07-13  5753861
898 2015-07-14  5059171
899 2015-07-15  4779650
900 2015-07-16  4535275
901 2015-07-17  4164812
902 2015-07-18  3157657
903 2015-07-19    55404
904 2015-07-20  3996292
905 2015-07-21  3515258
906 2015-07-22  3337937
907 2015-07-23  3594960
908 2015-07-24  3499777
909 2015-07-25  2847063
910 2015-07-26    55506
911 2015-07-27  5603824
912 2015-07-28  4910264
913 2015-07-29  4524391
914 2015-07-30  4685380
915 2015-07-31  5380356

[916 rows x 2 columns]

In [ ]:
dates =[]
for i in final_storeA_sales.Date.values:
    
    if i not in final_A_10pct_sales.Date.values:
        dates.append(i)
        print(i)
len(dates)


for i in range(len(date)):
    final_A_10pct_sales.loc[]

In [314]:
final_A_10pct_sales['Year'] = [i.year for i in final_A_10pct_sales.Date]

In [315]:
final_A_10pct_sales

Date        Sales  Year
0   2013-01-02  5804.266667  2013
1   2013-01-03  5332.116667  2013
2   2013-01-04  5431.300000  2013
3   2013-01-05  4363.100000  2013
4   2013-01-07  9426.950000  2013
5   2013-01-08  7579.650000  2013
6   2013-01-09  6420.500000  2013
7   2013-01-10  6552.900000  2013
8   2013-01-11  6633.650000  2013
9   2013-01-12  4385.000000  2013
10  2013-01-14  5030.816667  2013
11  2013-01-15  4742.066667  2013
12  2013-01-16  4462.050000  2013
13  2013-01-17  4648.550000  2013
14  2013-01-18  5057.016667  2013
15  2013-01-19  4407.700000  2013
16  2013-01-21  7686.650000  2013
17  2013-01-22  7169.183333  2013
18  2013-01-23  6578.666667  2013
19  2013-01-24  6201.883333  2013
20  2013-01-25  6353.616667  2013
21  2013-01-26  4235.300000  2013
22  2013-01-28  5060.366667  2013
23  2013-01-29  4628.950000  2013
24  2013-01-30  4952.133333  2013
25  2013-01-31  5928.500000  2013
26  2013-02-01  5891.116667  2013
27  2013-02-02  4872.683333  2013
28  2013-02-04  8841.283333  2013
29  2013-02-05  7609.150000  2013
..         ...          ...   ...
754 2015-06-27  4732.366667  2015
755 2015-06-29  9988.600000  2015
756 2015-06-30  9666.050000  2015
757 2015-07-01  7532.233333  2015
758 2015-07-02  6918.266667  2015
759 2015-07-03  6780.533333  2015
760 2015-07-04  4519.033333  2015
761 2015-07-06  6266.266667  2015
762 2015-07-07  5299.250000  2015
763 2015-07-08  5630.150000  2015
764 2015-07-09  5707.033333  2015
765 2015-07-10  5812.033333  2015
766 2015-07-11  4695.400000  2015
767 2015-07-13  8905.750000  2015
768 2015-07-14  7787.666667  2015
769 2015-07-15  7288.733333  2015
770 2015-07-16  6980.733333  2015
771 2015-07-17  6434.283333  2015
772 2015-07-18  4742.450000  2015
773 2015-07-20  6074.283333  2015
774 2015-07-21  5308.800000  2015
775 2015-07-22  5050.916667  2015
776 2015-07-23  5429.950000  2015
777 2015-07-24  5325.600000  2015
778 2015-07-25  4302.683333  2015
779 2015-07-27  8705.266667  2015
780 2015-07-28  7490.733333  2015
781 2015-07-29  6890.616667  2015
782 2015-07-30  7167.600000  2015
783 2015-07-31  8217.950000  2015

[784 rows x 3 columns]

In [318]:
final_A_10pct_sales_2014 = final_A_10pct_sales[final_A_10pct_sales['Year']==2014]

In [321]:
final_A_10pct_sales_2014['WeekofYear'] = [w.isocalendar()[1] for w in final_A_10pct_sales_2014.Date]

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [325]:
final_A_10pct_sales_2014_weekly = final_A_10pct_sales_2014.groupby(['WeekofYear']).mean()

In [326]:
final_A_10pct_sales_2014_weekly

Sales  Year
WeekofYear                   
1           5693.901263  2014
2           7229.460334  2014
3           4910.692090  2014
4           6573.011299  2014
5           5321.055556  2014
6           7220.277778  2014
7           5193.980556  2014
8           6673.686111  2014
9           5250.427778  2014
10          7076.165230  2014
11          5014.822222  2014
12          6541.802778  2014
13          5130.188889  2014
14          7245.272222  2014
15          5368.622222  2014
16          8425.760000  2014
17          5115.633333  2014
18          8026.720000  2014
19          7118.102778  2014
20          5106.925000  2014
21          6757.444444  2014
22          6040.971073  2014
23          7499.415254  2014
24          5656.555932  2014
25          7268.296045  2014
26          5232.559463  2014
27          7562.873990  2014
28          5225.563636  2014
29          6735.966667  2014
30          4973.603030  2014
31          7017.209091  2014
32          6619.381818  2014
33          5080.281818  2014
34          6766.500000  2014
35          5610.712121  2014
36          7039.508586  2014
37          5049.848765  2014
38          6482.802301  2014
39          4891.203030  2014
40          7498.072727  2014
41          6730.690909  2014
42          5005.618182  2014
43          6150.148485  2014
44          5710.632488  2014
45          7350.509091  2014
46          6589.112121  2014
47          5389.339141  2014
48          7803.866667  2014
49          8462.842424  2014
50          6093.687879  2014
51          9434.954545  2014
52          7422.345455  2014

In [332]:
final_storeA_sales

Date         Sales
0   2013-01-01   2907.000000
1   2013-01-02   6512.269103
2   2013-01-03   5870.261231
3   2013-01-04   6007.013311
4   2013-01-05   4940.978333
5   2013-01-06   2084.666667
6   2013-01-07  10173.510851
7   2013-01-08   8267.909850
8   2013-01-09   7174.076795
9   2013-01-10   7202.126878
10  2013-01-11   7328.373957
11  2013-01-12   5016.136895
12  2013-01-13   1395.000000
13  2013-01-14   5587.325543
14  2013-01-15   5282.098497
15  2013-01-16   5029.631052
16  2013-01-17   5129.268781
17  2013-01-18   5585.440735
18  2013-01-19   4932.218698
19  2013-01-20   1614.500000
20  2013-01-21   8248.333890
21  2013-01-22   7813.176962
22  2013-01-23   7168.213689
23  2013-01-24   6831.530885
24  2013-01-25   6998.978297
25  2013-01-26   4821.392321
26  2013-01-27   1068.333333
27  2013-01-28   5634.202003
28  2013-01-29   5154.100167
29  2013-01-30   5525.570952
..         ...           ...
886 2015-07-02   7595.953411
887 2015-07-03   7472.813644
888 2015-07-04   4912.855241
889 2015-07-05   3106.461538
890 2015-07-06   6883.539101
891 2015-07-07   5908.978369
892 2015-07-08   6206.572379
893 2015-07-09   6215.760399
894 2015-07-10   6282.113145
895 2015-07-11   5117.290000
896 2015-07-12   3598.916667
897 2015-07-13   9589.768333
898 2015-07-14   8431.951667
899 2015-07-15   7966.083333
900 2015-07-16   7571.410684
901 2015-07-17   6952.941569
902 2015-07-18   5262.761667
903 2015-07-19   4617.000000
904 2015-07-20   6660.486667
905 2015-07-21   5858.763333
906 2015-07-22   5563.228333
907 2015-07-23   5991.600000
908 2015-07-24   5832.961667
909 2015-07-25   4745.105000
910 2015-07-26   4625.500000
911 2015-07-27   9339.706667
912 2015-07-28   8183.773333
913 2015-07-29   7540.651667
914 2015-07-30   7808.966667
915 2015-07-31   8967.260000

[916 rows x 2 columns]

---


In [334]:
final_storeA_sales.head()

Date        Sales
0 2013-01-01  2907.000000
1 2013-01-02  6512.269103
2 2013-01-03  5870.261231
3 2013-01-04  6007.013311
4 2013-01-05  4940.978333

In [335]:
final_storeA_sales['Year'] = [i.year for i in final_storeA_sales.Date]
final_storeA_sales_2014 = final_storeA_sales[final_storeA_sales['Year']==2014]
final_storeA_sales_2014['WeekofYear'] = [w.isocalendar()[1] for w in final_storeA_sales_2014.Date]
final_storeA_sales_2014_weekly = final_storeA_sales_2014.groupby(['WeekofYear']).mean()

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [336]:
final_storeA_sales_2014_weekly.head()

Sales  Year
WeekofYear                   
1           5443.828368  2014
2           7996.732592  2014
3           5403.401423  2014
4           7208.619049  2014
5           5572.298075  2014

In [337]:
from bokeh.plotting import figure, output_file, show

output_file("2014_weekly_sales.html")

p = figure(plot_width=1000, plot_height=700)

# add a line renderer
p.line(final_storeA_sales_2014_weekly.index, final_storeA_sales_2014_weekly['Sales'], line_width=2,line_color = 'red')
p.line(final_A_10pct_sales_2014_weekly.index,final_A_10pct_sales_2014_weekly['Sales'], line_width=2,line_color = 'blue')
show(p)

In [340]:
sc_y.inverse_transform(y_test)

array([[3821.],
       [4732.],
       [7439.],
       ...,
       [5174.],
       [4944.],
       [6338.]])

In [341]:
len(y_test)==len(y_pred_svm)

True

In [344]:
l1 = [a[0] for a in y_test]

In [346]:
len(l1)

9197

In [349]:
output_file("predictions.html")

p2 = figure(plot_width=1000, plot_height=700)

# add a line renderer
p2.circle(np.arange(len(y_test)), [a[0] for a in sc_y.inverse_transform(y_test)], line_width=2,line_color = 'red')
p2.circle(np.arange(len(y_pred_svm)),[b[0] for b in sc_y.inverse_transform(y_pred_svm)], line_width=2,line_color = 'blue')
show(p2)